03-pipefilter.md

```diff
@@ -16,277 +16,465 @@ Now that we know a few basic commands,
+> When we tell the shell to run a program,
+> it creates a new process
+> and temporarily sends whatever we type on our keyboard to that process's standard input,
+> and whatever the process sends to standard output to the screen.
+>
+> Here's what happens when we run `ls *.pdb > content-of-molecules`.
+> The shell starts by telling the computer to create a new process to run the `ls` program.
+> `ls` reads from standard input, there is nothing to read.
+> And since we've used `>` to redirect output to a file,
+> the shell connects the process's standard output to that file.
+>
+> If we run `ls *.pdb | wc -l` instead,
+> the shell creates two processes
+> (one for each process in the pipe)
+> so that `ls` and `wc` run simultaneously.
+> The standard output of `ls` is fed directly to the standard input of `wc`;
+> since there's no redirection with `>`,
+> `wc`'s output goes to the screen.
+>
+> ![Redirects and Pipes](fig/redirects-and-pipes.png)
+>
+> This simple idea is why Unix has been so successful.
+> Instead of creating enormous programs that try to do many different things,
+> Unix programmers focus on creating lots of simple tools that each do one job well,
+> and that work well with each other.
+> This programming model is called "pipes and filters".
+> We've already seen pipes;
+> a **filter** is a program like `wc`
+> that transforms a stream of input into a stream of output.
+> Almost all of the standard Unix tools can work this way:
+> unless told to do otherwise,
+> they read from standard input,
+> do something with what they've read,
+> and write to standard output.
+>
+> The key is that any program that reads lines of text from standard input
+> and writes lines of text to standard output
+> can be combined with every other program that behaves this way as well.
+> You can *and should* write your programs this way
+> so that you and other people can put those programs into pipes to multiply their power.
+
+A second question that we can ask is
+"how many files inside the `molecules` directory are alcohols?".
+If you remember a little of chemistry,
+the name of alcohols ended with "nol".
+Let's run the command `ls *nol.pdb`:
 
 ~~~ {.bash}
-$ sort -n lengths.txt
+$ ls *nol.pdb
 ~~~
 ~~~ {.output}
-  9  methane.pdb
- 12  ethane.pdb
- 15  propane.pdb
- 20  cubane.pdb
- 21  pentane.pdb
- 30  octane.pdb
-107  total
+cyclohexanol.pdb
+ethanol.pdb
+methanol.pdb
 ~~~
 
-We can put the sorted list of lines in another temporary file called `sorted-lengths.txt`
-by putting `> sorted-lengths.txt` after the command,
-just as we used `> lengths.txt` to put the output of `wc` into `lengths.txt`.
-Once we've done that,
-we can run another command called `head` to get the first few lines in `sorted-lengths.txt`:
+The `*` in `*nol.pdb` matches zero or more characters,
+so the shell turns `*nol.pdb` into a complete list
+of files that end with `nol.pdf`
+
+> ## Wildcards {.callout}
```

Comment:
I would move this back to the top, to line 30/77 - we use the wildcard at the very start, so we should explain what it is then too.  One thought: the first paragraph could be outside the callout (as it's necessary information) and the rest of the callout can remain, but be entitled "more about wildcards"